torch/distributed/elastic/rendezvous/dynamic_rendezvous.py

```diff
@@ -0,0 +1,299 @@
+            )
+
+        if max_nodes < min_nodes:
+            raise ValueError(
+                f"The maximum number of nodes ({max_nodes}) must be greater than or equal to the "
+                f"minimum number of nodes ({min_nodes})."
+            )
+
+        self._run_id = run_id
+
+        self._store = store
+        self._backend = backend
+
+        self._min_nodes = min_nodes
+        self._max_nodes = max_nodes
+
+        self._timeout = timeout or RendezvousTimeout()
+
+    @property
+    def run_id(self) -> str:
+        """Gets the run id of the rendezvous."""
+        return self._run_id
+
+    @property
+    def store(self) -> Store:
+        """Gets the C10d store returned as part of the rendezvous."""
+        return self._store
+
+    @property
+    def backend(self) -> RendezvousBackend:
+        """Gets the backend used to hold the rendezvous state."""
+        return self._backend
+
+    @property
+    def min_nodes(self) -> int:
+        """Gets the minimum number of nodes to admit to the rendezvous."""
+        return self._min_nodes
+
+    @property
+    def max_nodes(self) -> int:
+        """Gets the maximum number of nodes to admit to the rendezvous."""
+        return self._max_nodes
+
+    @property
+    def timeout(self) -> RendezvousTimeout:
+        """Gets the timeout configuration of the rendezvous."""
+        return self._timeout
+
+    def get_backend(self) -> str:
+        """See base class."""
+        return self._backend.name
+
+    def next_rendezvous(self) -> Tuple[Store, int, int]:
+        """See base class."""
+        raise NotImplementedError()
+
+    def is_closed(self) -> bool:
+        """See base class."""
+        raise NotImplementedError()
+
+    def set_closed(self) -> None:
+        """See base class."""
+        raise NotImplementedError()
+
+    def num_nodes_waiting(self) -> int:
+        """See base class."""
+        raise NotImplementedError()
+
+    def get_run_id(self) -> str:
+        """See base class."""
+        return self._run_id
+
+    def shutdown(self) -> bool:
+        """See base class."""
+        raise NotImplementedError()
+
+
+def _get_timeout(params: RendezvousParameters, key: str) -> Optional[timedelta]:
+    timeout = params.get_as_int(key + "_timeout")
+    if timeout is None:
+        return None
+    return timedelta(seconds=timeout)
+
+
+def create_handler(
+    store: Store, backend: RendezvousBackend, params: RendezvousParameters
+) -> DynamicRendezvousHandler:
+    """Create a new :py:class:`DynamicRendezvousHandler` from the specified
+    parameters.
+
+    +-------------------+------------------------------------------------------+
+    | Parameter         | Description                                          |
+    +===================+======================================================+
+    | join_timeout      | The total time, in seconds, within which the         |
```

Comment:
Do you think maybe it makes sense to use a single timeout for join_timeout, last_call_timeout and close_timeout ?
I wonder when users would need to explicitly set each one separately.

Reply:
I think in most of the cases (if at all) the user would be just adjusting `join_timeout` only. Combining these timeouts into one does not make much sense in my opinion as they are used for different purposes. `close_timeout` is used for `set_closed()` and `shutdown()` and should ideally be much shorter than `join_timeout`. `last_call_timeout` on the other hand is used for early completion of the rendezvous and is mostly independent of `join_timeout`. `join_timeout` more or less determines how long a call to `next_rendezvous()` should take.